syntax = "proto3";


message B4EPayload{

  enum Action {
    CREATE_ACTOR = 0;
    CREATE_INSTITUTION = 1;
    CREATE_TEACHER = 2;
    CREATE_EDU_PROGRAM = 3;
    CREATE_CLASS = 4;
    CREATE_VOTING = 5;
    VOTE = 6;
    CREATE_RECORD = 7;
    CREATE_CERT = 8;
    CREATE_SUBJECT = 9;
    UPDATE_RECORD = 10;
    MODIFY_SUBJECT = 11;
    MODIFY_CERT = 12;
    REVOKE_CERT = 13;
    REACTIVE_CERT = 14;
    UPDATE_ACTOR_PROFILE = 15;
    REJECT_INSTITUTION = 16;
    ACTIVE_INSTITUTION = 17;

    SET_B4E_ENVIRONMENT = 18;
  }

  Action action = 1;

  CreateActorAction create_actor = 2;
  CreateActorAction create_institution = 3;
  CreateTeacherAction create_teacher = 4;
  CreatePortfolioAction create_edu_program = 5;
  CreateClassAction create_class = 6;
  CreateVotingAction create_voting = 7;
  VoteAction vote = 8;
  CreateRecordAction create_record = 9;
  CreateRecordAction create_cert = 10;
  CreateRecordAction create_subject = 11;
  UpdateRecordAction update_record = 12;
  ModifyRecordAction modify_subject = 13;
  ModifyRecordAction modify_cert = 14;
  ChangeStatusCertAction revoke_cert = 15;
  ChangeStatusCertAction reactive_cert = 16;
  UpdateActorProfileAction update_actor_profile = 17;
  RejectInstitutionAction reject_institution = 18;
  ActiveInstitutionAction active_institution = 19;
  SetB4EEnvironmentAction set_b4e_environment = 20;

  uint64 timestamp = 21;
}

message CreateActorAction {
  string id = 1;
  string data = 2;
}
message CreateTeacherAction {
  string id = 1;
  string teacher_public_key = 2;
  string data = 3;
}

enum PortfolioType{
  EDU_PROGRAM = 0;
}
message CreatePortfolioAction{
  string id = 1;
  string owner_public_key = 2;
  PortfolioType portfolio_type = 3;
  string data = 4;
}

message CreateClassAction{
  string class_id = 1;
  string subject_id = 2;
  int32 credit = 3;
  string teacher_public_key = 4;
  repeated string student_public_keys = 5;
}

enum VoteType{
  ACTIVE = 0;
  REJECT = 1;
}
message CreateVotingAction{
  string elector_public_key = 1;
  VoteType vote_type = 2;
}

message VoteAction {
  string elector_public_key = 1;
  bool accept = 2;
}

enum RecordType{
  SUBJECT = 0;
  CERTIFICATE = 1;
  OTHER = 2;
}
message CreateRecordAction {
  string owner_public_key = 1;
  string manager_public_key = 2;
  string record_id = 3;
  RecordType record_type = 4;
  string portfolio_id = 5;
  string cipher = 6;
  string hash = 7;
}

enum RecordStatus{
  CREATED = 0 ;
  REVOKED = 1;
  REACTIVATED = 2;
}

message UpdateRecordAction {
  string owner_public_key = 1;
  string record_id = 3;
  string cipher = 4;
  string hash = 5;
  RecordStatus record_status = 6;
}

message ModifyRecordAction{
  string owner_public_key = 1;
  string manager_public_key = 2;
  string record_id = 3;
  string cipher = 4;
  string hash = 5;
}
message ChangeStatusCertAction{
  string owner_public_key = 1;
  string record_id = 2;

}

message UpdateActorProfileAction {
  string data = 1;
}

message  RejectInstitutionAction{
  string institution_public_key = 1;
}

message  ActiveInstitutionAction{
  string institution_public_key = 1;
}

message SetB4EEnvironmentAction{
  uint64 timestamp = 1;
}


